Nats-top package addition

AnmolVirdi opened the pull request:
**Nats-top package** `Current version: v.0.6.3`
- top-like tool for monitoring NATS servers.

**Type**: Go binary
**REF**: https://github.com/nats-io/nats-top

**Tests:**
- nats-top, nats server and 2 client mode
- delimiter, refresh delays, subscriptions, sort restuls
- TLS Mode

AnmolVirdi committed:
Added melange config for nats-top with tests

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Fixed epoch number

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Merge branch 'main' into nats-top-adder

AnmolVirdi committed:
Fixed linting issue

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Merge branch 'main' into nats-top-adder

mamccorm commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+
+  - uses: go/build
+    with:
+      packages: ./nats-top.go
```
We should drop this and have it default to building go.mod, which imports nats-top.go:
 - https://github.com/nats-io/nats-top/blob/main/go.mod

Otherwise we may lose important dependencies which are included in go.mod

AnmolVirdi replied:
Noted, I've updated default to building go.mod

AnmolVirdi committed:
Setting default build to go.mod

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Merge branch 'wolfi-dev:main' into nats-top-adder

AnmolVirdi committed:
Updated tests

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Merge branch 'main' into nats-top-adder

AnmolVirdi committed:
Merge branch 'main' into nats-top-adder

AnmolVirdi committed:
Improved tests, added validations

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi commented:
- Improved quality of tests.
- Added more validations.
- Replaced `openssl` with `mkcert` for certs creation for test in TLS mode. But I'm still using openssl in the test environment to verify these certs.

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+        nats-top --version 2>&1 | grep ${{package.version}}
+    - name: Test nats-top with minimal nats server and 2 clients
+      runs: |
+        nats-server -p 4222 -m 8222 & sleep 2 &
```
Can we move it into a dedicated step, named something like "Run nats-server"?

AnmolVirdi replied:
Yes, noted. Making changes accordingly

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+    - name: Test nats-top with minimal nats server and 2 clients
+      runs: |
+        nats-server -p 4222 -m 8222 & sleep 2 &
+        nats sub client1 > /dev/null & sleep 2 & nats sub client2 > /dev/null & sleep 2
```
Can we move these subscriber runs into a separate step?

AnmolVirdi replied:
I'm adding nats server and nats sub together in a separate step since both of them are needed to set up server environment for `nats-top` tests

maxgio92 replied:
Also with the server is fine, I'd just remove it from the tests as it's part of the fixtures setup

AnmolVirdi replied:
I'm not sure if moving it out of tests will work since we're re-setting up nats server later in the TLS mode test

maxgio92 replied:
I tested the below locally, which works
```yml
    - name: Run NATS server
      runs: |
        nats-server -p 4222 -m 8222 & sleep 2 &
    - name: Run NATS subscribers
      runs: |
        nats sub client2 > /dev/null &
        sleep 2
        nats sub client1 > /dev/null &
        sleep 2
    - name: Test correct clients are in nats-top output
      runs: |
        nats-top -s localhost -b -u -o output.txt
        grep -q client1 output.txt
        grep -q client2 output.txt
        grep -q "Connections Polled: 2" output.txt
[...]
```

AnmolVirdi replied:
I've added changes accordingly. Thanks

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+        nats-server -p 4222 -m 8222 & sleep 2 &
+        nats sub client1 > /dev/null & sleep 2 & nats sub client2 > /dev/null & sleep 2
+        nats-top -s localhost -b -u -o nats-output-${{package.version}}.txt
+        cat nats-output-${{package.version}}.txt
```
This will never fail. What's the expected output from nats-top?

I would check that the output contains the exact client names and that connections polled is 2 in the output.

Something like:
```yml
    - name: Test correct clients are in nats-top output
      runs: |
        nats-top -s localhost -b -u -o output.txt
        grep -q client1 output.txt
        grep -q client2 output.txt
        grep -q "Connections Polled: 2" output.txt
```

AnmolVirdi replied:
Thanks for pointing this, making changes accordingly.

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+        cat nats-output-${{package.version}}.txt
+    - name: test connection limit (-n) and sorting options (-sort)
+      runs: |
+        nats-top -n 1 -sort uptime -o nats-output-${{package.version}}.txt
```
I would test those options separately as with one connection the sorting result wouldn't change.

Something like

```yml
    - name: test connection limit (-n)
      runs: |
        nats-top -n 1 -o output.txt
        grep -q "Connections Polled: 1" output.txt
    - name: test sorting options (-sort)
      runs: |
        nats-top -sort uptime -u -o output.txt
        grep -q client2 <(tail -1 output.txt)
```

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+        mkcert -install
+        mkcert -cert-file server-cert.pem -key-file server-key.pem localhost 127.0.0.1
+        mkcert -client -cert-file client-cert.pem -key-file client-key.pem localhost 127.0.0.1
+        cp "$(mkcert -CAROOT)/rootCA.pem" ca-cert.pem
```
Can we move the PKI preparation into a separate step?

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,91 @@
+        pkill nats-server & sleep 2
+        # Certs validation before injecting them into NATS
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/server-cert.pem
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/client-cert.pem
```
Can we move the PKI preparation into a separate step? (all of the above commands, basically, until this one included).
I think we can skip the certificate validation, as we're basically testing `mkcert` by doing so, which is out of scope from this test suite I would say.

AnmolVirdi replied:
- I tried adding certs validation to another step, but somehow the certs generated were not found in the next step. Even though I used docker runner

- I believe if in future TLS mode issue occurs, it could be due to 2 reasons: Nats-top tls issue or certs issue. Having a separate verification step would allow us to discover the issue quicker, i.e., if step fails, it's only because of nats-top tls issue and never certs since they're validated.

maxgio92 replied:
> I tried adding certs validation to another step, but somehow the certs generated were not found in the next step. Even though I used docker runner
Yeah we need to move them out of /tmp

> I believe if in future TLS mode issue occurs, it could be due to 2 reasons: Nats-top tls issue or certs issue. Having a separate verification step would allow us to discover the issue quicker, i.e., if step fails, it's only because of nats-top tls issue and never certs since they're validated.

I see your point, makes sense. It's harmless anyway, fine for me.

AnmolVirdi replied:
Thanks for investigating it was `tmp` dir issue. The tests work fine now!

AnmolVirdi committed:
Improved steps

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,98 @@
+          echo "Error: CSV File is empty!"
+          exit 1
+        fi
+        head -n 5 "nats-output-${{package.version}}.csv"
```
`head` doesn't exit 1 when the number of lines are greater than the ones of the file being read.

Instead, I'd validate via Python `csv` module to make it easier, like
```yml
    - name: test output with delimiter (-l)
      runs: |
        nats-top -o output.csv -l ","
        if [ ! -s "output.csv" ]; then
          echo "Error: CSV File is empty!"
          exit 1
        fi
        # Sanitize the input
        sed -i '/^$/d' output.csv
        # Ensure at least one column for each rows
        python -c "import csv; reader = csv.reader(open('output.csv', 'r')); [print(row[0]) for row in reader]"
```

AnmolVirdi replied:
Thanks for pointing this, i've updated the test.

AnmolVirdi committed:
Fix delimmiter test for nats-top

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,98 @@
+        head -n 5 "nats-output-${{package.version}}.csv"
+    - name: test setting refresh delay (-d) and refresh count (-r)
+      runs: |
+        export TERM=xterm && timeout 3 script -q -c "nats-top -d 1 -r 3" /dev/null
```
Is it required to set $TERM and `script`?
It seems that with `script` doesn't propagate the exit code:

```sh
/home/build # export TERM=xterm && timeout 3 script -c "nats-top -d 1 -non-existent-flag 3"
Script started, output log file is 'typescript'.
flag provided but not defined: -non-existent-flag

usage: nats-top [-s server] [-m http_port] [-ms https_port] [-n num_connections] [-d delay_secs] [-r max] [-o FILE] [-l DELIMITER] [-sort by]
                [-cert FILE] [-key FILE] [-cacert FILE] [-k] [-b] [-v|--version] [-u|--display-subscriptions-column]

Script done.
/home/build # echo $?
0
/home/build # export TERM=xterm && timeout 3 nats-top -d 1 -non-existent-flag 3
flag provided but not defined: -non-existent-flag

usage: nats-top [-s server] [-m http_port] [-ms https_port] [-n num_connections] [-d delay_secs] [-r max] [-o FILE] [-l DELIMITER] [-sort by]
                [-cert FILE] [-key FILE] [-cacert FILE] [-k] [-b] [-v|--version] [-u|--display-subscriptions-column]

/home/build # echo $?
1
```

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+        python -c "import csv; reader = csv.reader(open('nats-output-${{package.version}}.csv', 'r')); [print(row[0]) for row in reader]"
+    - name: test setting refresh delay (-d) and refresh count (-r)
+      runs: |
+        export TERM=xterm && timeout 3 script -q -c "nats-top -d 1 -r 3" /dev/null
```
```suggestion
        export TERM=xterm
        script --quiet --return --command "timeout 3 nats-top -d 1 -r 3"
```
`--return` propagates the exit code from timeout that will propagate the exit code from `nats-top`, so test will fail when expected

AnmolVirdi replied:
Thank you for suggesting this! We can use `script` for running any interactive tests now, for future cases.

maxgio92 replied:
Yes, that's awesome indeed @AnmolVirdi, thanks

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+        nats-server -p 4223 -m 8223 --tls --tlscert ./tmp/certs/server-cert.pem --tlskey ./tmp/certs/server-key.pem --tlscacert ./tmp/certs/ca-cert.pem & sleep 2
+        # Creating a NATS client and publishing sample message
+        nats sub -s "localhost:4223" --tlscert ./tmp/certs/client-cert.pem --tlskey ./tmp/certs/client-key.pem --tlsca ./tmp/certs/ca-cert.pem "test.subject" > /dev/null & sleep 1
+        nats-top -s localhost -m 8223 -cert ./tmp/certs/client-cert.pem -key ./tmp/certs/client-key.pem -cacert ./tmp/certs/ca-cert.pem -o nats-output-${{package.version}}.txt
```
Surprisingly it seems that `-cert`/`-key`/`-cacert` are no-op:

```shell
/home/build # nats-top -s localhost -m 8223 -cert pippo -key pluto -cacert paperino -o /dev/null
/home/build # echo $?
0
```

maxgio92 replied:
Reproduced also with upstream via Go, on a Ubuntu container, so it doesn't seem our issue:
```shell
docker run --rm -it --entrypoint bash ubuntu:latest
$ apt update && apt install -y golang >/dev/null 2>&1
$ go install github.com/nats-io/nats-server/v2@main
$ go install github.com/nats-io/natscli/nats@latest
$ go install github.com/nats-io/nats-top@latest
# generate certs
$ nats-server -p 4223 -m 8223 --tls --tlscert server-cert.pem --tlskey server-key.pem --tlscacert ca-cert.pem >/dev/null &
$ nats sub -s "localhost:4223" --tlscert client-cert.pem --tlskey client-key.pem --tlsca ca-cert.pem "test.subject" >/dev/null &
$ nats-top -s localhost -m 8223 -cert non-existent-cert.txt -key non-existent-key.pem -cacert non-existent-ca-cert.pem -o /dev/null
$ echo $?
0
```

AnmolVirdi replied:
Thanks for pointing this out @maxgio92.
This is because in the current test we were accidentally trying to access http_port and not `https`. I've pushed another commit which makes sure the certs are being used properly.
- Added `tlsverify` flag to make sure we use certs for client verifications
- Accessing nats-top using tls port

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+        # Certs validation before injecting them into NATS
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/server-cert.pem
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/client-cert.pem
+        # Setting up NATS server
+        nats-server -p 4223 -m 8223 --tls --tlscert ./tmp/certs/server-cert.pem --tlskey ./tmp/certs/server-key.pem --tlscacert ./tmp/certs/ca-cert.pem & sleep 2
+        # Creating a NATS client and publishing sample message
+        nats sub -s "localhost:4223" --tlscert ./tmp/certs/client-cert.pem --tlskey ./tmp/certs/client-key.pem --tlsca ./tmp/certs/ca-cert.pem "test.subject" > /dev/null & sleep 1
+        nats-top -s localhost -m 8223 -cert ./tmp/certs/client-cert.pem -key ./tmp/certs/client-key.pem -cacert ./tmp/certs/ca-cert.pem -o nats-output-${{package.version}}.txt
+        grep -q "Connections Polled: 1" nats-output-${{package.version}}.txt >/dev/null 2>&1
```
```suggestion
    - name: Run nats server in TLS mode
      runs: |
        # Killing previous nats-server process
        pkill nats-server & sleep 2
        # Setting up NATS server
        nats-server -p 4223 -m 8223 --tls --tlscert ./certs/server-cert.pem --tlskey ./certs/server-key.pem --tlscacert ./certs/ca-cert.pem & sleep 2
    - name: Run NATS subscribers in TLS mode
      runs: |
        nats sub -s "localhost:4223" --tlscert ./certs/client-cert.pem --tlskey ./certs/client-key.pem --tlsca ./certs/ca-cert.pem "test.subject" > /dev/null & sleep 1
    - name: Test NATS top in TLS mode
      runs: |
        nats-top -s localhost -m 8223 -cert ./certs/server-cert.txt -key ./certs/client-key.pem -cacert ./certs/ca-cert.pem -o output.txt
        grep -q "Connections Polled: 1" output.txt
```

AnmolVirdi replied:
I've added these changes in the next commit.

maxgio92 commented:
Sorry for being pedantic, once addressed all the point this test suite is going to be very good! Thanks!

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+        nats-top -s localhost -b -u -o nats-output-${{package.version}}.txt
+        grep -q "client1" nats-output-${{package.version}}.txt >/dev/null 2>&1
+        grep -q "client2" nats-output-${{package.version}}.txt >/dev/null 2>&1
+        grep -q "Connections Polled: 2" nats-output-${{package.version}}.txt >/dev/null 2>&1
```
`>/dev/null 2>&1` not required with `grep -q`

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+    - name: test connection limit (-n)
+      runs: |
+        nats-top -n 1 -o nats-output-${{package.version}}.txt
+        grep -q "Connections Polled: 1" nats-output-${{package.version}}.txt >/dev/null 2>&1
```
`>/dev/null 2>&1` not required with `grep -q`

AnmolVirdi replied:
Removed, Thanks

AnmolVirdi commented on nats-top.yaml:
```diff
@@ -0,0 +1,102 @@
+    - name: test setting refresh delay (-d) and refresh count (-r)
+      runs: |
+        export TERM=xterm && timeout 3 script -q -c "nats-top -d 1 -r 3" /dev/null
+    - name: Test Nats-top in TLS mode
+      runs: |
+        mkdir -p /tmp/certs
+        cd /tmp/certs
+        # Using mkcert to generate certs
+        mkcert -install
+        mkcert -cert-file server-cert.pem -key-file server-key.pem localhost 127.0.0.1
+        mkcert -client -cert-file client-cert.pem -key-file client-key.pem localhost 127.0.0.1
+        cp "$(mkcert -CAROOT)/rootCA.pem" ca-cert.pem
+        cd ../..
+        # Killing previous nats-server process
+        pkill nats-server & sleep 2
+        # Certs validation before injecting them into NATS
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/server-cert.pem
+        openssl verify -CAfile ./tmp/certs/ca-cert.pem ./tmp/certs/client-cert.pem
```
- Removing the tmp directory and moving certs generation to separate step.

AnmolVirdi committed:
Improved TLS tests

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Fixed linting issue

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Fixed linting issue

Signed-off-by: Anmol Virdi <anmol.virdi@chainguard.dev>

AnmolVirdi committed:
Merge branch 'main' into nats-top-adder

maxgio92 commented on nats-top.yaml:
```diff
@@ -0,0 +1,112 @@
+        sleep 1
+    - name: Test NATS top in TLS mode
+      runs: |
+        nats-top -s "localhost" -ms 8223 -u -cert ./certs/client-cert.pem -key ./certs/client-key.pem -cacert ./certs/ca-cert.pem -o nats-output-${{package.version}}.txt
```
@AnmolVirdi  So as discussed it seems that we need to explicitely set the URL to use TLS (https) otherwise -cert/-key/-cacert are no op (it strange nats-top doesn't complain BTW)

AnmolVirdi replied:
The issue was with passing tls mode via `-ms` flag for https port. Earlier I was using '-m' which is default http port.
